### ISSUE TITLE
Think about interaction with clap to allow loading/overriding config values from CLI args

### ISSUE THREAD

**Issue by LukasKalbertodt** [LukasKalbertodt/confique#19]
Often it's useful to allow users to override some (or all) configuration values via command line. Confique should work nicely for that use case. The currently best way to do it is probably to convert the CLI values to a partial type (manually) and then add it via `Builder::preloaded`. I would like to investigate whether this can be made more convenient and with less duplicate code. 

If this improvement has to be CLI-library specific, I am pretty sure I only want to support clap. I only ever use clap with the derive feature and I think it's the most mature library. 

Just to throw some random ideas into this issue, maybe one can annotate config fields with `#[config(clap(...))]` and if any are annotated this way, we will generate an additional type containing the fields annotated that way that has the `derive(clap::*)` on it. And has a method to convert it to a partial config type. This extra type can then be flattened into the main clap type. But again, haven't thought about this too deeply yet.

**Comment by ssokolow** [comment]
For the record, I also use [Gumdrop](https://lib.rs/crates/gumdrop) in my actix-web projects, where I don't need the path arguments to support accepting non-UTF8-able POSIX paths (i.e. where they're just going to be a path to a server root folder or config file) and I want to benefit from the reduced compile times and smaller output binaries.

**Comment by LukasKalbertodt** [comment]
Fair point. It's certainly worth considering how one could support multiple cli libraries. Or even be completely library-agnostic.

**Comment by hmuendel** [comment]
I would also love to have an integration with clap. I followed your workaround described above and I find it quite ok. However I was not able to derive `Config` directly on my clap cli flag struct, because of defaults. I totally agree with your view, that the finally parsed config should not contain options any longer, but this lead me to have options in my clap struct to see if the user provided the flag. I expect the flag to take precedence over all other config options (since it is the most direct user input). So I had another struct with options removed and config derived on it, which I could then construct from the partials. 
But this misses out on the documentation benefits from clap.

I didn't see a better way of doing it for now, but I still find your crate awesome and clean and can live with the solution. Just wanted to raise the point of defaults in clap and confique.

**Comment by Nemo157** [comment]
A different kind of approach to support overriding via the command line is to have a general flag that accepts a fragment of config and applies it, this is the approach taken by `cargo --config` and `jj --config-toml`. This seems relatively simple to do now with `confique`, and is maybe just worth adding an example of, I used something like:

```rust
#[derive(confique::Config, Debug)]
#[config(partial_attr(derive(Clone, Debug)))]
#[config(partial_attr(serde(deny_unknown_fields)))]
struct Config {
    ...
}

type Partial = <Config as confique::Config>::Partial;

impl Config {
    fn load(fragments: Vec<Partial>) -> Result<Self> {
        let dirs = ProjectDirs::from(...).ok_or_eyre("cannot get config directory")?;
        let mut builder = Config::builder();
        // reverse so that later fragments take precedence
        for fragment in fragments.into_iter().rev() {
            builder = builder.preloaded(fragment)
        }
        Ok(builder.env().file(dirs.config_dir().join("config.toml")).load()?)
    }
}

#[derive(Debug, Parser)]
struct App {
    /// Config overrides to apply, these should be fragments of the config file.
    #[arg(long = "config-toml", value_name = "TOML", value_parser = toml::from_str::<Partial>)]
    config_fragments: Vec<config::Partial>,

    ...
}

fn main() -> Result<()> {
    let app = App::parse();
    let config = Config::load(app.config_fragments)?;
    tracing::trace!(?config, "loaded config");

    ...
}
```

**Comment by aschey** [comment]
I think this can be accomplished without having to add a dependency on `clap` or have any implementations specific to it. We need to be able to use `partial_attr` on struct fields (PR created for this: #44):

```rust
use clap::Args;

#[derive(confique::Config, Debug)]
#[config(partial_attr(derive(Args, Debug)))]
pub struct AppConfig {
    #[config(partial_attr(arg(long)))]
    port: usize,
    #[config(partial_attr(arg(long)))]
    allowed_hosts: Vec<String>,
}
```

which allows you to embed the partial struct inside a clap `Parser`:

```rust
use clap::Parser;
struct Cli {
    #[command(flatten)]
    app_config: PartialAppConfig,
    // other fields
}
```

and use it like this:

```rust
let cli = Cli::parse();
let config = AppConfig::builder()
    .preloaded(cli.app_config)
    .env()
    .file(&config_path)
    .load()?;

```

The only remaining issue is that clap requires structs to use `Option` without the fully-qualified name, but the derive macro used here generates structs that use `std::option::Option`, causing clap's parsing to fail. My initial reaction was that this should be fixed in clap, but apparently this is a [feature](https://docs.rs/clap/latest/clap/_derive/index.html#arg-types) to work around the lack of true compile-time reflection.

To get around this, we could either
- provide another attribute to the derive macro to allow users to override the exact name used when generating optional fields
- remove the fully-qualified path and generate structs that use `Option` instead of `std::option::Option`

I think the second option is _probably_ fine since `Option` is in the global prelude, but I could be convinced otherwise.

**Comment by LukasKalbertodt** [comment]
Regarding the fully qualified `Option`: as far as I can see it, either we use the fully qualified path or we fully disallow a symbol `Option` to refer to anything else than `std::option::Option` in the `derive(Config)` scope. Which is prooobably fine, in particular since `confique` also attributes special behavior to `Option` fields. There are some theoretical counter examples, where this would be a breaking change. But mh, I'm not entirely opposed to changing this.

<details>

Just changing `confique` to emit `Option` instead of `std::option::Option` would make this break:

```rs
#[derive(Config)]
struct MealConfig {
    option: Option,
}

enum Option { Vegan, Vegetarian, Anything }
```

Even if no type called `Option` is used in the config struct, but is just in scope, it is complicated. The inner module (with the partial type) has `use super::*;` because the user might use types for struct fields that are defined next to the main `Config` struct. These need to be imported in the partial module. And the macro cannot know where a field type is defined, so cannot know if it needs to be imported.

</details>

But isn't it possible to use `std::option::Option` with clap and still get the behavior by specifying the appropriate attributes (`.action(ArgAction::Set).required(false)`)?

**Comment by LukasKalbertodt** [comment]
I just published v0.4.0 which includes support for `layer_attr` on struct fields, uses plain `Option` in generated code (meaning that clap understands the type), and forward docs to the layer type. With this, I think there is decent support for CLI stuff via clap for example. You don't need to duplicate docs or flag definitions. I still think many apps would opt for a separate type for CLI args. 

I'd be interested in getting some feedback on this from people who actually use confique and use CLI as one layer to load the config. Is the solution as is fine? What are pain points?

Once I gathered a bit more experience, I will again update the docs to give nicer guidance on possible ways to integrate CLI with confique.

**Comment by LukasKalbertodt** [comment]
Oh and take a look at the `clap` example I added: https://github.com/LukasKalbertodt/confique/blob/main/examples/clap.rs
